net, init: derive default onion port if a user specified a -port

mzumsande opened the pull request:
This resolves #31133 (setups with multiple local nodes each using a different `-port` no longer working with v28.0, see the issue description for more details) by deriving the default onion listening port to be the value specified by `-port` incremented by 1 (idea by vasild / laanwj).
Note that with this fix, the chosen `-port` values of two local nodes cannot be adjacent, otherwise there will be port collisions again.

From the discussion in the linked issue, this was the most popular option, followed by doing nothing and telling affected users to change their setups to use `-bind` instead of `-port`. But more opinions are certainly welcome!

I think that if we decide to do something about the problem described in the issue, we should do so soon (in 28.1.), so I opened this PR.
Fixes #31133

mzumsande committed:
net, init: derive default onion port if a user specified a -port

After port collisions are no longer tolerated but lead to
a startup failure in v28.0, local setups of multiple nodes,
each with a different -port value would not be possible anymore
due to collision of the onion default port - even if the nodes
were using tor or not interested in receiving onion inbound connections.

Fix this by deriving the onion listening port to be -port + 1.
(idea by vasild / laanwj)

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

mzumsande committed:
Add release note for #31223

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

Christewart commented:
concept ACK

tdb3 reviewed (commented):
Approach ACK

This seems like a reasonable and straightforward way to alleviate the issue.

Seemed to work nicely with two node instances run as follows:
```
build/src/bitcoind -port=11000 -rpcport=11100 -datadir=/mnt/tmp/n1/
```
```
build/src/bitcoind -port=12000 -rpcport=12100 -datadir=/mnt/tmp/n2/
```

```
ss -tulpn
...

... 127.0.0.1:11100                    0.0.0.0:*           users:(("bitcoind",pid=192480,fd=12))       
... 127.0.0.1:11001                    0.0.0.0:*           users:(("bitcoind",pid=192480,fd=29))       
... 127.0.0.1:12100                    0.0.0.0:*           users:(("bitcoind",pid=192511,fd=12))       
... 127.0.0.1:12001                    0.0.0.0:*           users:(("bitcoind",pid=192511,fd=29))                                                         
... 0.0.0.0:11000                    0.0.0.0:*           users:(("bitcoind",pid=192480,fd=32))       
... 0.0.0.0:12000                    0.0.0.0:*           users:(("bitcoind",pid=192511,fd=32))

...
```

laanwj commented:
Concept ACK

mzumsande commented on src/torcontrol.cpp:
```diff
@@ -711,9 +711,9 @@ void StopTorControl()
     struct in_addr onion_service_target;
     onion_service_target.s_addr = htonl(INADDR_LOOPBACK);
-    return {onion_service_target, BaseParams().OnionServiceTargetPort()};
+    return {onion_service_target, port.has_value() ? port.value() : BaseParams().OnionServiceTargetPort()};
```
Thanks, changed to that.

vasild commented on src/init.cpp:
```diff
@@ -1907,7 +1907,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                InitWarning(BadPortWarning("-port", *onion_port));
+            }
+        }
+        onion_service_target = DefaultOnionServiceTarget(onion_port);
```
I think it is simpler and is ok to use reduce all this to:

```diff
diff --git i/src/init.cpp w/src/init.cpp
index c10f792c7d..7e141dbd4c 100644
--- i/src/init.cpp
+++ w/src/init.cpp
@@ -518,13 +518,13 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
                  " If <type> is not supplied or if <type> = 1, indexes for all known types are enabled.",
                  ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
 
     argsman.AddArg("-addnode=<ip>", strprintf("Add a node to connect to and attempt to keep the connection open (see the addnode RPC help for more info). This option can be specified multiple times to add multiple nodes; connections are limited to %u at a time and are counted separately from the -maxconnections limit.", MAX_ADDNODE_CONNECTIONS), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
     argsman.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-bantime=<n>", strprintf("Default duration (in seconds) of manually configured bans (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-bind=<addr>[:<port>][=onion]", strprintf("Bind to given address and always listen on it (default: 0.0.0.0). Use [host]:port notation for IPv6. Append =onion to tag any incoming connections to that address and port as incoming Tor connections (default: 127.0.0.1:%u=onion, testnet3: 127.0.0.1:%u=onion, testnet4: 127.0.0.1:%u=onion, signet: 127.0.0.1:%u=onion, regtest: 127.0.0.1:%u=onion)", defaultBaseParams->OnionServiceTargetPort(), testnetBaseParams->OnionServiceTargetPort(), testnet4BaseParams->OnionServiceTargetPort(), signetBaseParams->OnionServiceTargetPort(), regtestBaseParams->OnionServiceTargetPort()), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-bind=<addr>[:<port>][=onion]", strprintf("Bind to given address and always listen on it (default: 0.0.0.0). Use [host]:port notation for IPv6. Append =onion to tag any incoming connections to that address and port as incoming Tor connections (default: 127.0.0.1:%u=onion, testnet3: 127.0.0.1:%u=onion, testnet4: 127.0.0.1:%u=onion, signet: 127.0.0.1:%u=onion, regtest: 127.0.0.1:%u=onion)", defaultChainParams->GetDefaultPort() + 1, testnetChainParams->GetDefaultPort() + 1, testnet4ChainParams->GetDefaultPort() + 1, signetChainParams->GetDefaultPort() + 1, regtestChainParams->GetDefaultPort() + 1), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
     argsman.AddArg("-cjdnsreachable", "If set, then this host is configured for CJDNS (connecting to fc00::/8 addresses would lead us to the CJDNS network, see doc/cjdns.md) (default: 0)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-connect=<ip>", "Connect only to the specified node; -noconnect disables automatic connections (the rules for this peer are the same as for -addnode). This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
     argsman.AddArg("-discover", "Discover own IP addresses (default: 1 when listening and no -externalip or -proxy)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-dns", strprintf("Allow DNS lookups for -addnode, -seednode and -connect (default: %u)", DEFAULT_NAME_LOOKUP), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-dnsseed", strprintf("Query for peer addresses via DNS lookup, if low on addresses (default: %u unless -connect used or -maxconnections=0)", DEFAULT_DNSSEED), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-externalip=<ip>", "Specify your own public address", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
@@ -1844,12 +1844,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     connOptions.whitelist_relay = args.GetBoolArg("-whitelistrelay", DEFAULT_WHITELISTRELAY);
 
     // Port to bind to if `-bind=addr` is provided without a `:port` suffix.
     const uint16_t default_bind_port =
         static_cast<uint16_t>(args.GetIntArg("-port", Params().GetDefaultPort()));
 
+    const uint16_t default_bind_port_onion = default_bind_port + 1;
+
     const auto BadPortWarning = [](const char* prefix, uint16_t port) {
         return strprintf(_("%s request to listen on port %u. This port is considered \"bad\" and "
                            "thus it is unlikely that any peer will connect to it. See "
                            "doc/p2p-bad-ports.md for details and a full list."),
                          prefix,
                          port);
@@ -1868,13 +1870,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
                 continue;
             }
         } else {
             const std::string network_type = bind_arg.substr(index + 1);
             if (network_type == "onion") {
                 const std::string truncated_bind_arg = bind_arg.substr(0, index);
-                bind_addr = Lookup(truncated_bind_arg, BaseParams().OnionServiceTargetPort(), false);
+                bind_addr = Lookup(truncated_bind_arg, default_bind_port_onion, false);
                 if (bind_addr.has_value()) {
                     connOptions.onion_binds.push_back(bind_addr.value());
                     continue;
                 }
             }
         }
@@ -1904,22 +1906,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     CService onion_service_target;
     if (!connOptions.onion_binds.empty()) {
         onion_service_target = connOptions.onion_binds.front();
     } else if (!connOptions.vBinds.empty()) {
         onion_service_target = connOptions.vBinds.front();
     } else {
-        std::optional<uint16_t> onion_port;
-        if (args.IsArgSet("-port")) {
-            // If -port was specified but no onion bind, derive the onion listening port by incrementing
-            // that port by 1. This avoids port collisions in case of setups with multiple local nodes.
-            onion_port = args.GetIntArg("-port", 0) + 1;
-            if (IsBadPort(*onion_port)) {
-                InitWarning(BadPortWarning("-port", *onion_port));
-            }
-        }
-        onion_service_target = DefaultOnionServiceTarget(onion_port);
+        onion_service_target = DefaultOnionServiceTarget(default_bind_port_onion);
         connOptions.onion_binds.push_back(onion_service_target);
     }
 
     if (args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION)) {
         if (connOptions.onion_binds.size() > 1) {
             InitWarning(strprintf(_("More than one onion bind address is provided. Using %s "
diff --git i/src/torcontrol.h w/src/torcontrol.h
index f57bf9e65f..0b66201cf1 100644
--- i/src/torcontrol.h
+++ w/src/torcontrol.h
@@ -24,13 +24,13 @@ extern const std::string DEFAULT_TOR_CONTROL;
 static const bool DEFAULT_LISTEN_ONION = true;
 
 void StartTorControl(CService onion_service_target);
 void InterruptTorControl();
 void StopTorControl();
 
-CService DefaultOnionServiceTarget(std::optional<uint16_t> port);
+CService DefaultOnionServiceTarget(uint16_t port);
 
 /** Reply from Tor, can be single or multi-line */
 class TorControlReply
 {
 public:
     TorControlReply() { Clear(); }
diff --git i/src/torcontrol.cpp w/src/torcontrol.cpp
index d56216a7cb..60df916f16 100644
--- i/src/torcontrol.cpp
+++ w/src/torcontrol.cpp
@@ -708,12 +708,12 @@ void StopTorControl()
         torControlThread.join();
         event_base_free(gBase);
         gBase = nullptr;
     }
 }
 
-CService DefaultOnionServiceTarget(std::optional<uint16_t> port)
+CService DefaultOnionServiceTarget(uint16_t port)
 {
     struct in_addr onion_service_target;
     onion_service_target.s_addr = htonl(INADDR_LOOPBACK);
-    return {onion_service_target, port.has_value() ? port.value() : BaseParams().OnionServiceTargetPort()};
+    return {onion_service_target, port};
 }
diff --git i/src/chainparamsbase.h w/src/chainparamsbase.h
index c75a70cb96..adbd6a5174 100644
--- i/src/chainparamsbase.h
+++ w/src/chainparamsbase.h
@@ -19,21 +19,19 @@ class ArgsManager;
  */
 class CBaseChainParams
 {
 public:
     const std::string& DataDir() const { return strDataDir; }
     uint16_t RPCPort() const { return m_rpc_port; }
-    uint16_t OnionServiceTargetPort() const { return m_onion_service_target_port; }
 
     CBaseChainParams() = delete;
-    CBaseChainParams(const std::string& data_dir, uint16_t rpc_port, uint16_t onion_service_target_port)
-        : m_rpc_port(rpc_port), m_onion_service_target_port(onion_service_target_port), strDataDir(data_dir) {}
+    CBaseChainParams(const std::string& data_dir, uint16_t rpc_port)
+        : m_rpc_port(rpc_port), strDataDir(data_dir) {}
 
 private:
     const uint16_t m_rpc_port;
-    const uint16_t m_onion_service_target_port;
     std::string strDataDir;
 };
 
 /**
  * Creates and returns a std::unique_ptr<CBaseChainParams> of the chosen chain.
  */
diff --git i/src/chainparamsbase.cpp w/src/chainparamsbase.cpp
index aadd04e509..060d519d92 100644
--- i/src/chainparamsbase.cpp
+++ w/src/chainparamsbase.cpp
@@ -38,21 +38,21 @@ const CBaseChainParams& BaseParams()
  * been chosen arbitrarily to keep ranges of used ports tight.
  */
 std::unique_ptr<CBaseChainParams> CreateBaseChainParams(const ChainType chain)
 {
     switch (chain) {
     case ChainType::MAIN:
-        return std::make_unique<CBaseChainParams>("", 8332, 8334);
+        return std::make_unique<CBaseChainParams>("", 8332);
     case ChainType::TESTNET:
-        return std::make_unique<CBaseChainParams>("testnet3", 18332, 18334);
+        return std::make_unique<CBaseChainParams>("testnet3", 18332);
     case ChainType::TESTNET4:
-        return std::make_unique<CBaseChainParams>("testnet4", 48332, 48334);
+        return std::make_unique<CBaseChainParams>("testnet4", 48332);
     case ChainType::SIGNET:
-        return std::make_unique<CBaseChainParams>("signet", 38332, 38334);
+        return std::make_unique<CBaseChainParams>("signet", 38332);
     case ChainType::REGTEST:
-        return std::make_unique<CBaseChainParams>("regtest", 18443, 18445);
+        return std::make_unique<CBaseChainParams>("regtest", 18443);
     }
     assert(false);
 }
 
 void SelectBaseParams(const ChainType chain)
 {
```

which will also handle variants like `-port=5555 -bind=127.0.0.1=onion` (will result in listening on `127.0.0.1:5556` for incoming Tor connections).

mzumsande replied:
took the suggestion now!

vasild reviewed (commented):
Another alternative: don't listen for Tor on regtest by default (were all reports for this for regtest?).

What about setups that use e.g. `-port=5000` and have a manually configured the Tor hidden service in `torrc` that redirects to `127.0.0.1:8334`? E.g.:

```
HiddenServiceDir /var/db/tor/myservice/
HiddenServiceVersion 3
HiddenServicePort 8333 127.0.0.1:8334
```

This change would _silently_ break those setups - because we will now be listening on `127.0.0.1:5001` and other peers that try to connect to the node will get a failure because the Tor router gets connection refused on `127.0.0.1:8334` and this may remain unnoticed for a long time.

i-am-yuvi commented:
Concept ACK

mzumsande reviewed (commented):
>Another alternative: don't listen for Tor on regtest by default (were all reports for this for regtest?).

I don't know, but I could see people / orgs having local setups with multiple nodes on mainnet, e.g. with one node facing random peers from the internet/allowing inbounds, and others that only connect to local trusted peers. In general, I don't like regtest-specific rules so I would see them as a last resort if nothing else works.

> What about setups that use e.g. -port=5000 and have a manually configured the Tor hidden service in torrc that redirects to 127.0.0.1:8334?

Good point, I agree that this is a downside of the port +1 approach. I'm really not sure how widespread this kind of behavior is, at least our tor.md doesn't recomment using `-port` for anything related to tor.

Haven't pushed any changes yet, will do a bit more testing for vasild's suggested change / also look into a functional test next week.

laanwj commented:
> Another alternative: don't listen for Tor on regtest by default (were all reports for this for regtest?).

Agree with @mzumsande in that i don't particularly like this; remember that regtest is primarily for testing mainnet behavior in our functional tests, so we want the behavior to be as close to that as possible. Every extra flag needed makes for more test-only code.

vasild commented on doc/release-notes-31223.md:
```diff
@@ -0,0 +1,6 @@
+-----------------------
+When the `-port` configuration option is used, the default onion listening port will now
+be derived to be that port + 1 instead of being set to a fixed value.
+This re-allows setups with multiple local nodes using different `-port`, which
+would lead to a startup failure in v28.0 due to a port collision error. (#31223)
```
I meant `-bind=` as a means to disable the onion bind on `127.0.0.1:8334` to avoid the port collision, but I typed `-bind=...=onion` instead :(

One can use both depending on what's needed: `-bind=1.2.3.4:5555 -bind=127.0.0.1:8335=onion`.

vasild commented on src/init.cpp:
```diff
@@ -1907,7 +1907,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        if (args.IsArgSet("-port")) {
+            // If -port was specified but no onion bind, derive the onion listening port by incrementing
+            // that port by 1. This avoids port collisions in case of setups with multiple local nodes.
+            onion_port = args.GetIntArg("-port", 0) + 1;
+            if (IsBadPort(*onion_port)) {
+                InitWarning(BadPortWarning("-port", *onion_port));
+            }
```
> Do you think that is an issue we should care about?

No.

I have found simple/dumb programs easier to configure compared to programs that try to be intelligent on behalf of the user. 

> Shaw's Principle:
>        Build a system that even a fool can use, and only a fool will want to use it.

This is a bit subjective and I may have a bias here, but if I put `Port 80` in `/etc/ssh/sshd_config` I can't imagine sshd saying something like "Hey, you are binding sshd on port 80, your web server may not work well". :-D

vasild commented:
> It's important to make this clear in the release notes.

Note that the problem this PR is trying to resolve has a release note and an easy workaround. We are here because some people (many?) didn't read the release notes even after encountering the problem (second bitcoind refuses to start because of port collision). A startup failure is an error that cannot be ignored. Whereas a silent breakage of the hidden service could remain unnoticed for a long time - one would have to check the number of incoming Tor connections, wonder why they are 0 over long period of time and investigate... I just wonder if the cure (this PR) can turn out to be worse than the disease.

mzumsande commented:
Updated PR (took much longer than expected due to a medical issue, sorry) and I still plan on adding a functional test, hopefully tomorrow.

- took vasild's suggestion, so that OnionServiceTargetPort is removed from `chainparamsbase`
- expanded release note

vasild commented on doc/release-notes-31223.md:
```diff
@@ -0,0 +1,8 @@
+be derived to be that port + 1 instead of being set to a fixed value (8334 on mainnet).
+This re-allows setups with multiple local nodes using different `-port` and not using `-bind`,
+which would lead to a startup failure in v28.0 due to a port collision.
+Note that a `HiddenServicePort` manually configured in `torrc` may need adjustment if used in
+connection with the `-port` option. (#31223)
```
Maybe further elaborate this with an example:

If you are using `-port=` with non-standard value, for example `-port=5555` and not using `-bind=...=onion`, previously Bitcoin Core would listen for incoming Tor connections on `127.0.0.1:8334`. Now it would listen on `127.0.0.1:5556` (`-port` plus one). If you configured the hidden service manually in `torrc` now you have to change it from `HiddenServicePort 8333 127.0.0.1:8334` to `HiddenServicePort 8333 127.0.0.1:5556`, or configure `bitcoind` with `-bind=127.0.0.1:8334=onion` to get the previous behavior.

vasild reviewed (commented):
I reviewed the code and it looks safe and will achieve the intended purpose.

Since I introduced the current behavior which this PR is aiming to fix, I understand that I may have a bias towards downplaying the current issue (port collision) and a biased preference towards "It's fine, don't do anything". However, even with this realization, I can't help but think whether this will [introduce a bigger problem than the one it solves](https://github.com/bitcoin/bitcoin/pull/31223#issuecomment-2470680219): silent breakage vs obvious breakage.

Thus I will neither approve nor disapprove this. In other words - I do not object this if enough other people think it is good.

mzumsande committed:
test: add functional test for -port behavior

mzumsande commented:
Updated Release note and added a commit that adds a functional test for various aspects of `-port` that have been discussed in this PR an the issue.

When it comes to the question whether the issue should be fixed in 28.1, I guess it also comes down to how widespread we think a scenario with `-port` and a manual `torrc` is, vs how many more people doing local testing setups would be affected by the `-port` change from 28.0.
I can't say I have a good feeling for either - to be honest I would have expected way more people to complain about their local setups not working anymore, than actually did in the last couple of weeks. 

But, to be clear, if reviewers think it's better not to make another change now and tell affected users to  use `-bind` instead of `-port`, I'm happy to close this / maybe still add a functional test for `-port` so that future changes in this area are more likely to be caught by tests.

tdb3 reviewed (approved):
Code review ACK 1dd3af8fbc350c6f1efa8ae6449e67e1b42ccff4

> if reviewers think it's better not to make another change now and tell affected users to use -bind instead of -port

I don't feel very strongly either way, but I lean toward implementing the port+1 approach in this PR.

> still add a functional test for -port so that future changes in this area are more likely to be caught by tests.

I support this. It's descriptive and helps catch issues later.

achow101 commented:
ACK 1dd3af8fbc350c6f1efa8ae6449e67e1b42ccff4

laanwj commented on src/chainparamsbase.h:
```diff
@@ -22,15 +22,13 @@ class CBaseChainParams
 public:
     const std::string& DataDir() const { return strDataDir; }
     uint16_t RPCPort() const { return m_rpc_port; }
-    uint16_t OnionServiceTargetPort() const { return m_onion_service_target_port; }
```
i think we could have kept this, and made it return `port + 1`? It would have made the overall patch smaller.

Oh wait, no, we don't have the normal P2P port information here. Never mind.

laanwj commented:
Tested ACK 1dd3af8fbc350c6f1efa8ae6449e67e1b42ccff4

fanquake commented:
Backported in #31469.
